edb/edgeql/compiler/inference/multiplicity.py

```diff
@@ -0,0 +1,701 @@
+) -> qltypes.Multiplicity:
+    # Here for debugging purposes.
+    raise ValueError('invalid infer_multiplicity(None, schema) call')
+
+
+@_infer_multiplicity.register
+def __infer_statement(
+    ir: irast.Statement,
+    *,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+    return infer_multiplicity(
+        ir.expr, scope_tree=scope_tree, ctx=ctx)
+
+
+@_infer_multiplicity.register
+def __infer_empty_set(
+    ir: irast.EmptySet,
+    *,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+    return ZERO
+
+
+@_infer_multiplicity.register
+def __infer_type_introspection(
+    ir: irast.TypeIntrospection,
+    *,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+    # The result is always ONE, but we still want to actually
+    # introspect the expression.
+    return ONE
+
+
+def _infer_shape(
+    ir: irast.Set,
+    *,
+    is_mutation: bool=False,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> None:
+    for shape_set, _ in ir.shape:
+        new_scope = cardinality._get_set_scope(shape_set, scope_tree)
+        if shape_set.expr and shape_set.rptr:
+            expr_mult = infer_multiplicity(
+                shape_set.expr, scope_tree=new_scope, ctx=ctx)
+
+            ptrref = shape_set.rptr.ptrref
+            if (expr_mult is MANY and
+                    not ptrref.out_target.is_scalar and
+                    not ptrref.out_target.collection):
+                raise errors.QueryError(
+                    f'possibly not a strict set returned by an '
+                    f'expression for a computable '
+                    f'{ptrref.shortname.name}.',
+                    hint=(
+                        f'Use DISTINCT for the entire computable expression '
+                        f'to resolve this.'
+                    ),
+                    context=shape_set.context
+                )
+
+        _infer_shape(
+            shape_set, is_mutation=is_mutation, scope_tree=scope_tree, ctx=ctx)
+
+
+def _infer_set(
+    ir: irast.Set,
+    *,
+    is_mutation: bool=False,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+    result = _infer_set_inner(
+        ir, is_mutation=is_mutation, scope_tree=scope_tree, ctx=ctx)
+    ctx.inferred_multiplicity[ir, scope_tree] = result
+    # The shape doesn't affect multiplicity, but requires validation.
+    _infer_shape(ir, is_mutation=is_mutation, scope_tree=scope_tree, ctx=ctx)
+
+    return result
+
+
+def _infer_set_inner(
+    ir: irast.Set,
+    *,
+    is_mutation: bool=False,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+    rptr = ir.rptr
+    new_scope = cardinality._get_set_scope(ir, scope_tree)
+
+    if rptr is not None:
+        # Validate the source regardless
```

Comment:
Regardless of what?

Reply:
Yeah, that's old comment (and the line is duplicated, too). Fixed.